### ISSUE TITLE
`vsphere-clone`: Packer stuck on "Waiting for IP" after setting `mac_address`

### ISSUE THREAD

**Issue by rconde01** [hashicorp/packer-plugin-vsphere#75]
#### Overview of the Issue
https://github.com/hashicorp/packer-plugin-vsphere/pull/73 fixes the fact that manual mac_address was still configured as automatic, but I still have the issue that packer can't connect when setting a manual mac address.

#### Reproduction Steps

Set a manual mac_address and packer gets stuck:
```
2021-06-14T13:41:05-04:00: ==> vsphere-clone.default: Cloning VM...
2021-06-14T13:41:07-04:00: ==> vsphere-clone.default: Mounting ISO images...
2021-06-14T13:41:07-04:00: ==> vsphere-clone.default: Adding configuration parameters...
2021-06-14T13:41:07-04:00: ==> vsphere-clone.default: Power on VM...
2021-06-14T13:41:08-04:00: ==> vsphere-clone.default: Waiting for IP...
```

### Plugin and Packer version

Packer 1.7.2 plugin 1.0.0

### Simplified Packer Buildfile

```
locals {
    vsphere_insecure_connection = "true"
}

packer {
  required_plugins {
    vsphere = {
      version = ">= 1.0.0"
      source  = "github.com/hashicorp/vsphere"
    }
  }
}

source "vsphere-clone" "default" {
    vcenter_server = var.vcenter_server
    username = var.vcenter_user
    password = var.vcenter_password
    insecure_connection = local.vsphere_insecure_connection

    datacenter = var.vcenter_datacenter
    datastore = var.vcenter_datastore
    cluster = var.vcenter_cluster
    host = var.vcenter_host

    folder = var.vcenter_dest_folder
    template = var.prev_template
    vm_name = var.new_template

    ssh_username = var.ssh_root_username
    ssh_password = var.ssh_root_password

    linked_clone = true
    create_snapshot = true

    network = var.vcenter_network
    mac_address = "00:50:56:83:3A:5D"
}

build {
    sources = [
        "source.vsphere-clone.default"
    ]
}
```

### Operating system and Environment details
Packer host: Windows 10
VM OS: Redhat EL7

### Log Fragments and crash.log files

```
Build packer images
2021/06/14 13:54:31 [INFO] Packer version: 1.7.2 [go1.16.3 windows amd64]
2021/06/14 13:54:31 [TRACE] discovering plugins in C:\ProgramData\chocolatey\lib\packer\tools
2021/06/14 13:54:31 [TRACE] discovering plugins in C:\Users\rob.conde\AppData\Roaming\packer.d\plugins
2021/06/14 13:54:31 [TRACE] discovering plugins in .
2021/06/14 13:54:31 [INFO] PACKER_CONFIG env var not set; checking the default config file path
2021/06/14 13:54:31 [INFO] PACKER_CONFIG env var set; attempting to open config file: C:\Users\rob.conde\AppData\Roaming\packer.config
2021/06/14 13:54:31 [WARN] Config file doesn't exist: C:\Users\rob.conde\AppData\Roaming\packer.config
2021/06/14 13:54:31 [INFO] Setting cache directory: C:\Users\rob.conde\Documents\XYZ\packer_cache
e: cannot determine if process is in background: Process background check error: not implemented yet
2021/06/14 13:54:31 [TRACE] listing potential installations for "github.com/hashicorp/vsphere" that match ">= 1.0.0". plugingetter.ListInstallationsOptions{FromFolders:[]string{"C:\\ProgramData\\chocolatey\\lib\\packer\\tools\\packer.exe", ".", "C:\\Users\\rob.conde\\AppData\\Roaming\\packer.d\\plugins"}, BinaryInstallationOptions:plugingetter.BinaryInstallationOptions{APIVersionMajor:"5", APIVersionMinor:"0", OS:"windows", ARCH:"amd64", Ext:".exe", Checksummers:[]plugingetter.Checksummer{plugingetter.Checksummer{Type:"sha256", Hash:(*sha256.digest)(0xc000c5e280)}}}}2021/06/14 13:54:31 [TRACE] Found the following "github.com/hashicorp/vsphere" installations: [{C:\Users\rob.conde\AppData\Roaming\packer.d\plugins\github.com\hashicorp\vsphere\packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe v1.0.0}]
2021/06/14 13:54:31 [INFO] found external [clone iso] builders from vsphere plugin
2021/06/14 13:54:31 [INFO] found external [-packer-default-plugin-name- template] post-processors from vsphere plugin
2021/06/14 13:54:31 [TRACE] Starting external plugin C:\Users\rob.conde\AppData\Roaming\packer.d\plugins\github.com\hashicorp\vsphere\packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe start builder clone
2021/06/14 13:54:31 Starting plugin: C:\Users\rob.conde\AppData\Roaming\packer.d\plugins\github.com\hashicorp\vsphere\packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe []string{"C:\\Users\\rob.conde\\AppData\\Roaming\\packer.d\\plugins\\github.com\\hashicorp\\vsphere\\packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe", "start", "builder", "clone"}
2021/06/14 13:54:31 Waiting for RPC address for: C:\Users\rob.conde\AppData\Roaming\packer.d\plugins\github.com\hashicorp\vsphere\packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe
2021/06/14 13:54:31 packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe plugin: 2021/06/14 13:54:31 Plugin port range: [10000,25000]
2021/06/14 13:54:31 packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe plugin: 2021/06/14 13:54:31 Plugin address: tcp 127.0.0.1:10000
2021/06/14 13:54:31 packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe plugin: 2021/06/14 13:54:31 Waiting for connection...
2021/06/14 13:54:31 Received tcp RPC address for C:\Users\rob.conde\AppData\Roaming\packer.d\plugins\github.com\hashicorp\vsphere\packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe: addr is 
127.0.0.1:10000
2021/06/14 13:54:31 packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe plugin: 2021/06/14 13:54:31 Serving a plugin connection...
2021/06/14 13:54:31 packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe plugin: 2021/06/14 13:54:31 [TRACE] starting builder clone
2021/06/14 13:54:31 Build debug mode: false
2021/06/14 13:54:31 Force build: false
2021/06/14 13:54:31 On error: abort
2021/06/14 13:54:31 Waiting on builds to complete...
2021/06/14 13:54:31 Starting build run: vsphere-clone.default
2021/06/14 13:54:31 Running builder:
2021/06/14 13:54:31 [INFO] (telemetry) Starting builder
vsphere-clone.default: output will be in this color.

2021/06/14 13:54:32 packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe plugin: 2021/06/14 13:54:32 No CD files specified. CD disk will not be made.
2021-06-14T13:54:32-04:00: ==> vsphere-clone.default: Cloning VM...
2021-06-14T13:54:34-04:00: ==> vsphere-clone.default: Mounting ISO images...2021/06/14 13:54:34 packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe plugin: 2021/06/14 13:54:34 No floppy files specified. Floppy disk will not be made.

2021-06-14T13:54:34-04:00: ==> vsphere-clone.default: Adding configuration parameters...
2021-06-14T13:54:34-04:00: ==> vsphere-clone.default: Power on VM...
2021/06/14 13:54:35 packer-plugin-vsphere_v1.0.0_x5.0_windows_amd64.exe plugin: 2021/06/14 13:54:35 [INFO] Waiting for IP, up to total timeout: 30m0s, settle timeout: 5s
2021-06-14T13:54:35-04:00: ==> vsphere-clone.default: Waiting for IP...
```

**Comment by rconde01** [comment]
I actually seem to be having the same issue with vsphere-iso and a linux vm...so maybe it's a linux issue?

**Comment by sylviamoss** [comment]
Are you able to see in the UI if the machine is stuck at some configuration step? This usually happens when a boot_command a pressed config is missing something 🤔  
So when my build gets stuck at this point I usually check the machine UI to check that.

**Comment by rconde01** [comment]
No...not that simple unfortunately. I can see in the vsphere interface that it does have an IP. If i remove the custom mac, it works fine. 

I'm not a "go" person, but I can see the relevant pieces:
https://github.com/hashicorp/packer-plugin-vsphere/blob/08e333e7db9f3fc6f72c4b6e165127f311359ec2/builder/vsphere/driver/vm.go#L640

https://github.com/vmware/govmomi/blob/9054158bb458f8c383422253709b56dc318ba2c0/object/virtual_machine.go#L320

Any chance to get logging in these functions to return more detail?

**Comment by rconde01** [comment]
Ok - I got my colleague to get me started building the plugin, logged the helI out of it, and found the issue (or at least 1 issue). The problem is in https://github.com/vmware/govmomi/blob/66b9538af589335e414a59ac8f046801947a64ff/object/virtual_machine.go#L336

When you get the mac address via nic.GetVirtualEthernetCard().MacAddress, it appears to preserve the case. But, down below 
https://github.com/vmware/govmomi/blob/66b9538af589335e414a59ac8f046801947a64ff/object/virtual_machine.go#L372

when it gets the mac address, it appears to always use lower case. It's looking for a literal match in the map so it fails. The loop never runs again and you just wait until it times out.

The workaround is to specify your mac address as lower case (assuming this works the same in all instances). For a fix, probably convert both to the same case.

So this is not in your library...can you run this down with vmware, or do you need me to file a bug over there? I guess you could convert what the user specifies to lower case but again that assumes that the mac coming from GuestNicInfo will always be lowercase...no idea if that's the case.

**Comment by rconde01** [comment]
This workaround works for vsphere-clone as well.

**Comment by sylviamoss** [comment]
Good catch, thanks for debugging! Maybe we could always make it lowercase as you said.
Sorry for the delay in replying, our team is working on a project and we've been focused on it. I'll analyze if making it lowercase is the best approach and implement it if so. 

I think you should open an issue directly to the govmomi repository. You could point to this issue and give them the exact same information you wrote here. It seems something interesting to fix. Also, you could probably open a PR over there suggesting the fix 🤔

**Comment by rconde01** [comment]
I will open an issue and PR with them...thx!

**Comment by rconde01** [comment]
Fix now merged, but not yet released:
https://github.com/vmware/govmomi/commit/0aa1de316a32e7a8cdf9683a02a40a3a4abf5300

**Comment by tenthirtyam** [comment]
- v1.0.7 updated to `vmware/govmomi@v0.26.1` in https://github.com/hashicorp/packer-plugin-vsphere/pull/176, which included the upstream fix. 

- v1.1.1 subsequently updated to `vmware/govmomi@v0.29.0` in https://github.com/hashicorp/packer-plugin-vsphere/pull/240.

**Comment by tenthirtyam** [comment]
Marking this issue as closed.

If this issue is reproducible with the [latest](https://github.com/hashicorp/packer-plugin-vsphere/releases/latest) version of the plugin please create a new issue linking back to this one for added context. 

Thank you!

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.
